text/0000-fabric-gateway.md

```diff
@@ -0,0 +1,213 @@
+  - The `Endorse` service is invoked on the Gateway, passing the ProposedTransaction message
+    - The Gateway will determine the endorsement plan for the requested chaincode and forward to the appropriate peers for endorsement. It will return to the client a `PreparedTransaction` message which contains a `Envelope` message as defined in `fabric-protos/common/common.proto`.  It will also contain other information, such as the return value of the chaincode function and the transaction ID so that the client doesn't necessarily need to unpack the `Envelope` payload.
+- __Submit__
+  - The client signs the hash of the Envelope payload using the user's private key and sets the signature field in the `Envelope` in the `PreparedTransaction`.
+  - The `Submit` service is invoked on the Gateway, passing the signed `PreparedTransaction` message.  A stream is opened to return multiple return values.
+    - The Gateway will register transaction event listeners for the given channel/txId.
+    - It will then broadcast the `Envelope` to the ordering service.
+    - The success/error response is passed back to the client in the stream
+    - The Gateway awaits suffient tx commit events before returning and closing the stream, indicating to the client that transaction has been committed.
+
+Evaluating a transaction is a simple process of invoking the `Evaluate` service passing a SignedProposal message.  The Gateway passes the request to a peer of it's choosing according to a defined policy (probably same org, highest block count) and returns the chaincode function return value to the client.
+
+## Launching the Gateway
+
+#### Standalone process
+When running standalone, the Gateway is effectively a 'client' application to a set of peers.  It will be configured to connect to a peer in the organization in order to invoke discovery.  To do this it will need a signing identity in the channel writers policy.
+
+To run the gateway server, the following parameters need to be supplied:
+- The url of at least one peer in the org.  Once connected, the discovery service will be invoked to find other peers.
+- The MSPID associated with the gateway.
+- The signing identity of the gateway (cert and key), e.g. location of PEM files on disk.
+- The TLS certificate so that the gateway can connect to other components in the organization
+The following is an example command line invocation using a prototype:
+- `gateway -h peer0.org1.example.com -p 7051 -m Org1MSP -id ../../fabric-samples/fabcar/javascript/wallet/gateway.id -tlscert ../../fabric-samples/test-network/organizations/peerOrganizations/org1.example.com/tlsca/tlsca.org1.example.com-cert.pem`
+
+#### Embedded in Peer
+When embedded in a peer, the gateway will register its gRPC service with the host peer's gRPC server.
+The gateway does not need its own (client) identity since it can get all the discovery information directly from the host peer rather than via a gRPC signed request.
```

Comment:
Running as part of the peer process, Fabric Gateway would have direct access to the local peer ledger updates / events and discovery information. It could track the ledger height of other peers by directly accessing gossip data. It wouldn't connect to other peer eventing services.